### PR TITLE
Batch updates and empty lists

### LINKED ISSUE #458: Allow for batch updates
The FirebaseUI adapters surface an atomic event for each update to a child item. In some cases it may be more efficient to update the UI only after a batch up updates has been received.

I'd like it if the FirebaseArray and adapters would raise an additional event when a `value` event is received. Given that 'value' events are guaranteed to fire after the corresponding `child_` events, this would be the perfect way to implement such an event.

    public void onDataChanged()

Note that we don't pass in a DataSnapshot here. The idea is that this is a flow-event only, you're supposed to capture the data in the regular child events and just just `onDataChanged` to capture the completion moment.

### PR DESCRIPTION
**Pull request by puf**
Fixes #458 and #118. Ready for a first review, but don't merge yet.

### CODE REVIEW

File: app/src/main/java/com/firebase/uidemo/database/ChatActivity.java [hunk 1]
```diff
@@ -74,23 +81,19 @@ protected void onCreate(Bundle savedInstanceState) {
...
-                            Log.e(TAG, "Failed to write message", databaseError.toException());
-                        }
-                    }
-                });
-
-                mMessageEdit.setText("");
+                sendCurrentMessage();
+            }
+        });
+        // Allow hitting the Send keyon the soft keyboard to send the message
```
Comment (morganchen12) [inline app/src/main/java/com/firebase/uidemo/database/ChatActivity.java]: nit: space between 'key' and 'on'

**Comment by SUPERCILEX** [comment]
Hey @puf,
I really love this PR, but **`onReady` will be inconsistently called when offline**. If a new ref is created offline and subscribed to via a `ValueEventListener` **`Query`** (such as `.orderByValue().equalTo(1)`), `onDataChanged` will never be called while offline whereas `onChildAdded` will. Because of this, **I don't think this PR should be merged until that issue is addressed because it could leave users waiting indefinitely for an `onReady` call that will never come**. I've already contacted Firebase support, but since you work on the database, I was hoping you could speed up the fix.

A mcve with plain old `ValueEventListener`/`ChildEventListener`s: https://github.com/SUPERCILEX/issue-mcve/blob/46f41526ab047f24241ac548b8ba6dd04b98cc42/app/src/main/java/com/supercilex/myapplication/MainActivity.java#L32
@puf If you would like, I can create a mcve with your PR to show you that `onReady` will not always be called when offline.

**Comment by puf** [comment]
Great find @SUPERCILEX . Agreed on holding off on merging until we figure this one out.

Unfortunately this is part of the behavior of the underlying SDK and not something we can change in FirebaseUI. It is not even something we can fix in the SDK, at least not without reworking a large part of the API surface. 

On the lower level we have no knowledge whether the ValueListener is listening for a single document (where we have a hard guarantee that we will never fire with incomplete results) or it is listening on a collection (where incomplete results are acceptable, as long as each individual child is complete).

Can you tell support to ping me on the case?

**Comment by SUPERCILEX** [comment]
@puf Yes, I told support to ping you.

@puf That's a real bummer for the internal api, but I still have hope because I've been using a hack to get around this problem in my app. In my mvce, if you ask the `ValueEventListener` for the parent database instead of the query once while online, that entire ref is cached and queries will work offline. From what you explained above, caching the parent ref guarantees complete results which is why it works.

Here's what I mean:
```java
// 1. While online, add a listener to the root ref:
mRef.addValueEventListener(this);
// 2. Go back to querying and it will work offline and online:
mRef.orderByValue().equalTo(1).addValueEventListener(this);
```

@puf Couldn't you guys do some sort of cache hack where if a query is executed offline and returns no results, you guys query the parent ref and use a for loop with `getChildren` to filter the results and return the proper query? If that doesn't work for the native Firebase apis, couldn't we use a similar hack for FirebaseArray?

**Comment by puf** [comment]
@SUPERCILEX: I'm against workarounds for behavior that is missing in the
Firebase SDK. Those types of workarounds tend to become a maintenance
problem as the SDK evolves.

**Comment by SUPERCILEX** [comment]
@puf sounds reasonable so we just wait for you guys to fix it on the SDK side.

**Comment by puf** [comment]
That is indeed my take. But this one likely won't be fixed soon, given that
it is inherent to the database's model: it's just a JSON tree. On the wire
we have no way to distinguish collections (where partial results are
acceptable and wanted) from "documents" (where we guarantee to never raise
partial results).

**Comment by samtstern** [comment]
I am going to start doing QA on version `1.1.0` as there are many important fixes in there and it's been longer than expected.  I have targeted this PR for `1.2.0` which will be the next minor release.

**Comment by samtstern** [comment]
@puf going to close this.  Here are the changes in this PR, as I see it:

- Batch events in the adapters.
- Soft keyboard and swipe support in the sample app.

We've already implemented the former (with your help) and the latter is really cool but not critical to the library.  Since this has been open for many months and is currently not mergeable there's not much to do with it.  If you ever get some extra time (unlikely, I know) and want to get the app changes merge-able again then I'm happy to make these changes.

**Pull request closed without merging**